streams/src/test/java/org/apache/kafka/streams/integration/IQv2IntegrationTest.java

```diff
@@ -0,0 +1,440 @@
+                IntegrationTestUtils.iqv2WaitForPartitionsOrGlobal(
+                    kafkaStreams,
+                    request,
+                    partitions
+                );
+
+            assertThat(result.getPartitionResults().keySet(), is(partitions));
+            for (final Integer partition : partitions) {
+                assertThat(result.getPartitionResults().get(partition).isFailure(), is(true));
+                assertThat(
+                    result.getPartitionResults().get(partition).getFailureReason(),
+                    is(FailureReason.NOT_UP_TO_BOUND)
+                );
+                assertThat(
+                    result.getPartitionResults().get(partition).getFailureMessage(),
+                    is("Query requires a running active task,"
+                        + " but partition was in state PARTITIONS_ASSIGNED and was active.")
+                );
+            }
+        }
+    }
+
+    @Test
+    public void shouldFetchFromPartition() {
+        final PingQuery query = new PingQuery();
+        final int partition = 1;
+        final Set<Integer> partitions = singleton(partition);
+        final StateQueryRequest<Boolean> request =
+            inStore(STORE_NAME).withQuery(query).withPartitions(partitions);
+
+        kafkaStreams.start();
+        final StateQueryResult<Boolean> result =
+            IntegrationTestUtils.iqv2WaitForPartitionsOrGlobal(kafkaStreams, request, partitions);
+
+        assertThat(result.getPartitionResults().keySet(), equalTo(partitions));
+    }
+
+    @Test
+    public void shouldFetchExplicitlyFromAllPartitions() {
+        final PingQuery query = new PingQuery();
+        final Set<Integer> partitions = mkSet(0, 1);
+        final StateQueryRequest<Boolean> request =
+            inStore(STORE_NAME).withQuery(query).withAllPartitions();
+
+        kafkaStreams.start();
+        final StateQueryResult<Boolean> result =
+            IntegrationTestUtils.iqv2WaitForPartitionsOrGlobal(kafkaStreams, request, partitions);
+
+        assertThat(result.getPartitionResults().keySet(), equalTo(partitions));
+    }
+
+    @Test
+    public void shouldNotRequireQueryHandler() {
+        final PingQuery query = new PingQuery();
+        final int partition = 1;
+        final Set<Integer> partitions = singleton(partition);
+        final StateQueryRequest<Boolean> request =
+            inStore(STORE_NAME).withQuery(query).withPartitions(partitions);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        builder.table(
+            INPUT_TOPIC_NAME,
+            Consumed.with(Serdes.Integer(), Serdes.Integer()),
+            Materialized.as(new KeyValueBytesStoreSupplier() {
+                @Override
+                public String name() {
+                    return STORE_NAME;
+                }
+
+                @Override
+                public KeyValueStore<Bytes, byte[]> get() {
+                    return new KeyValueStore<Bytes, byte[]>() {
```

Comment:
This takes up a lot of real estate, but what we're doing is creating a custom state store implementation that for sure doesn't implement the handler, which lets us verify the default behavior when stores don't implement the handler.

I didn't want to use any pre-existing store implementation for this, just in case we actually do decide to add a query handler to that store in the future.